### PR TITLE
feat: [CI-6709]: pipeline env variables in stage

### PR DESCRIPTION
**Pull request by devkimittal**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43025)

### CODE REVIEW

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java [hunk 1]
```diff
@@ -672,32 +707,38 @@ private ContainerResourceParams getIndependentContainerResource(
...
-  private Map<String, String> getEnvVariables(IntegrationStageNode stageNode) {
-    if (isEmpty(stageNode.getVariables())) {
-      return Collections.emptyMap();
+  private Map<String, String> getVariablesMap(List<NGVariable> variables, String stageId) {
+    Map<String, String> mapOfVariables = new HashMap<>();
+    if (isEmpty(variables)) {
+      return mapOfVariables;
+    }
+
+    for (NGVariable variable : variables) {
```
Comment (raghavharness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java:716]: Can you keep using stream for this ?

**Comment by harshjain12** [comment]
Can we add unit tests meanwhile i am reviewing

File: 332-ci-manager/service/src/main/java/io/harness/ci/beans/serializer/RunTimeInputHandler.java [hunk 2]
```diff
@@ -291,6 +291,40 @@ public static String resolveStringParameterWithDefaultValue(String fieldName, St
...
+    // It only checks input set pattern. Variable can be resolved on lite engine.
+    if (parameterField.isExpression() && matchesInputSetPattern(parameterField.getExpressionValue())) {
+      if (isMandatory && defaultValue == null) {
+        throw new CIStageExecutionUserException(
+            format("Failed to resolve mandatory field %s in step type %s with identifier %s", fieldName, stepType,
+                stepIdentifier));
+      } else {
+        log.warn(format("Failed to resolve optional field %s in step type %s with identifier %s", fieldName, stepType,
+            stepIdentifier));
+        return defaultValue.toString();
```
Comment (raghavharness) [inline 332-ci-manager/service/src/main/java/io/harness/ci/beans/serializer/RunTimeInputHandler.java:318]: should handle null here as well for non-mandatory cases.

**Review by raghavharness — CHANGES_REQUESTED** [review]

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java (hunk 1 above)
Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java:716]: Need to put two different conditions for String and number variable here, don't think that can be done using stream here.

Comment (raghavharness) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java:716]: You can use stream two times I think

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java [hunk 3]
```diff
@@ -347,13 +350,18 @@ private ContainerDefinitionInfo createPluginCompatibleStepContainerDefinition(Pl
...
+    envVarMap.putAll(getVariablesMap(stageNode.getPipelineVariables(), stageNode.getIdentifier()));
+    envVarMap.putAll(getVariablesMap(stageNode.getVariables(), stageNode.getIdentifier()));
     envVarMap.putAll(PluginSettingUtils.getBuildEnvironmentVariables(stepInfo, ciExecutionArgs));
     envVarMap.putAll(pluginSettingUtils.getPluginCompatibleEnvVariables(
         stepInfo, identifier, timeout, ambiance, StageInfraDetails.Type.K8, false));
     setEnvVariablesForHostedBuids(stageNode, stepInfo, envVarMap);
     Integer runAsUser = resolveIntegerParameter(stepInfo.getRunAsUser(), null);
 
+    Map<String, SecretNGVariable> secretVarMap = new HashMap<>();
+    secretVarMap.putAll(getSecretVariablesMap(stageNode.getPipelineVariables()));
```
[the remark this answers is not in the document]
Comment (devkimittal) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/integrationstage/K8InitializeStepUtils.java:362]: what if the both pipeline variable and stage variables are created with same name abc? stage one should override the pipeline one irrespectice of whether it is a account secret or a project secret, right?

**Review by raghavharness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 9 files at r1, 1 of 6 files at r2, 11 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43025#-:-NNHH9zO8vRhwbkKkyhw:bnfp4nl)* status: 1. unresolved discussions. Waiting on team(s): sto.

**Review by raghavharness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43025#-:-NNHPX2I7JhpIQd1YVe2:bnfp4nl)* status: 1. unresolved discussions. Waiting on team(s): sto.

**Review by raghavharness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43025)* status: 1. unresolved discussions. Waiting on team(s): sto.

**Pull request merged**